src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -228,8 +228,121 @@ ucs_status_t uct_rc_mlx5_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, size
     return status;
 }
 
-ucs_status_t uct_rc_mlx5_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
-                                     const void *payload, unsigned length)
+#if HAVE_IBV_EXP_DM
+typedef struct uct_rc_mlx5_dm_pack_data {
+    uint64_t   hdr;
+    size_t     length;
+    const void *payload;
+} uct_rc_mlx5_dm_pack_data_t;
+
+typedef union uct_rc_mlx5_dm_copy_data {
+    uct_rc_am_short_hdr_t am_hdr;
+    struct ibv_exp_tmh    tm_hdr;
+    uint64_t              data[2];
+    char                  tail[sizeof(uint64_t) * 2];
+} UCS_S_PACKED uct_rc_mlx5_dm_copy_data_t;
+
+static size_t uct_rc_mlx5_ep_am_short_dm_pack(void *dest, void *arg)
+{
+    uct_rc_mlx5_dm_pack_data_t *data = (uct_rc_mlx5_dm_pack_data_t*)arg;
+    uint64_t *dst = dest;
+    *dst = data->hdr;
+    memcpy(dst + 1, data->payload, data->length);
+    return sizeof(data->hdr) + data->length;
+}
+
+static void UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_ep_copy_to_dm(uct_rc_mlx5_dm_copy_data_t *cache, size_t hdr_len,
+                          const void *payload, size_t length, void *dm)
+{
+    size_t len          = sizeof(*cache) - hdr_len;
+    const uint64_t* src = payload + len;
+    uint64_t* dst       = dm;
+
+    ucs_assert(sizeof(*cache) >= hdr_len);
+
+    /* copy head of payload to tail of cache */
+    memcpy(cache->tail + hdr_len, payload, ucs_min(len, length));
+
+    UCS_STATIC_ASSERT(sizeof(*cache) <= sizeof(cache->data));
```

Comment:
This condition is always true, size of structure can't be less its field.

Reply:
Sorry, double miss :) for union it's fine

Reply:
right, this is union, and current implementation have 2-uint64_t limitation (may be extended in case)